x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/api/rules/bulk_actions/bulk_enable_disable_rules.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { RulesClient } from '@kbn/alerting-plugin/server';
+import { invariant } from '../../../../../../../common/utils/invariant';
+import type { PromisePoolError } from '../../../../../../utils/promise_pool';
+import type { MlAuthz } from '../../../../../machine_learning/authz';
+import type { RuleAlertType } from '../../../../rule_schema';
+import { validateBulkEnableRule } from '../../../logic/bulk_actions/validations';
+
+interface BulkEnableDisableRulesArgs {
+  rules: RuleAlertType[];
+  action: 'enable' | 'disable';
+  isDryRun?: boolean;
+  rulesClient: RulesClient;
+  mlAuthz: MlAuthz;
+}
+
+interface BulkEnableDisableRulesOutcome {
+  updatedRules: RuleAlertType[];
+  errors: Array<PromisePoolError<RuleAlertType, Error>>;
+}
+
+export const bulkEnableDisableRules = async ({
+  rules,
+  isDryRun,
+  rulesClient,
+  action: operation,
+  mlAuthz,
+}: BulkEnableDisableRulesArgs): Promise<BulkEnableDisableRulesOutcome> => {
+  const errors: Array<PromisePoolError<RuleAlertType, Error>> = [];
+  const updatedRules: RuleAlertType[] = [];
+
+  // In the first step, we validate if the rules can be enabled
+  const validatedRules: RuleAlertType[] = [];
+  await Promise.all(
+    rules.map(async (rule) => {
+      try {
+        await validateBulkEnableRule({ mlAuthz, rule });
+        validatedRules.push(rule);
+      } catch (error) {
+        errors.push({ item: rule, error });
+      }
+    })
+  );
+
+  if (isDryRun || validatedRules.length === 0) {
+    return {
+      updatedRules: validatedRules,
+      errors,
+    };
+  }
+
+  // Then if it's not a dry run, we enable the rules that passed the validation
+  const ruleIds = validatedRules.map(({ id }) => id);
+
+  // Perform actual update using the rulesClient
+  let results;
+  if (operation === 'enable') {
+    results = await rulesClient.bulkEnableRules({ ids: ruleIds });
+  } else {
+    results = await rulesClient.bulkDisableRules({ ids: ruleIds });
+  }
+  const failedRuleIds = results.errors.map(({ rule: { id } }) => id);
+
+  // Go through the original rules array and update rules that were not returned
+  // as failed from the bulkEnableRules. We cannot rely on the results from the
+  // bulkEnableRules because the response is not consistent and might not
```

Comment:
I'm curious what it means "the response isn't consistent"? Do you have examples of such inconsistent responses?

Reply:
Yes, some rules you pass to the bulk method might disappear from the response. I've expanded on this in the comment and included a link to a corresponding issue with examples: https://github.com/elastic/kibana/issues/181050.